File: client/android/div/src/main/java/com/yandex/div/core/util/mask/PhoneInputMask.kt
package com.yandex.div.core.util.mask

import com.yandex.div2.PhoneMasks
import org.json.JSONObject

internal class PhoneInputMask(
    private val onError: ((Exception) -> Unit)
) : BaseInputMask(DEFAULT_MASK_DATA) {

    override fun overrideRawValue(newRawValue: String) {
        tryInvalidateMaskDataWith(newRawValue)
        super.overrideRawValue(newRawValue)
    }

    override fun applyChangeFrom(newValue: String, position: Int?) {
        val textDiff = TextDiff.build(value, newValue).let { initialDiff ->
            if (position == null) return@let initialDiff
            TextDiff(
                start = (position - initialDiff.added).coerceAtLeast(0),
                added = initialDiff.added,
                removed = initialDiff.removed
            )
        }
        val oldRawValue = rawValue

        val tailStart = replaceBodyTail(textDiff, newValue)

        val newRawValue = rawValue
        val newPattern = newMaskPatternFor(newRawValue)

        if (newPattern == null) {
            calculateCursorPosition(textDiff, tailStart)
            return
        }

        updateMaskDataWith(newPattern)
        replaceChars(newRawValue, 0)

        val rawValueDiff = TextDiff.build(oldRawValue, newRawValue)
        val dynamicDestination = rawValueDiff.start + rawValueDiff.added
        calculateCursorPositionBy(dynamicDestination)
    }

    private fun calculateCursorPositionBy(dynamicDestination: Int) {
        var index = 0
        var dynamicCounter = 0

        while (index < destructedValue.size && dynamicCounter < dynamicDestination) {
            if (destructedValue[index++] is MaskChar.Dynamic) dynamicCounter++
        }

        cursorPosition = firstHolderAfter(index)
    }

    private fun tryInvalidateMaskDataWith(rawValue: String) =
        newMaskPatternFor(rawValue)?.let { newPattern -> updateMaskDataWith(newPattern) }

    private fun newMaskPatternFor(rawValue: String): String? {
        val newPattern = rawValue.phoneMaskPattern
        val currentPattern = maskData.pattern
        return if (newPattern != currentPattern) newPattern else null
    }

    private fun updateMaskDataWith(newPattern: String) = updateMaskData(
        newMaskData = MaskData(
            pattern = newPattern,
            decoding = DEFAULT_DECODING_MASK_KEY,
            alwaysVisible = maskData.alwaysVisible
        ),
        restoreValue = false
    )

    override fun onException(exception: Exception) {
        onError.invoke(exception)
    }
}

internal val DEFAULT_DECODING_MASK_KEY = listOf(
    BaseInputMask.MaskKey(
        key = '0',
        filter = "\\d",
        placeholder = '_'
    )
)

internal val DEFAULT_MASK_DATA = BaseInputMask.MaskData(
    pattern = "".phoneMaskPattern,
    decoding = DEFAULT_DECODING_MASK_KEY,
    alwaysVisible = false
)

private const val COUNTRY_CODE_END_MARKER = "*"
private const val UNIVERSAL_MASK = "000000000000000"

internal val String.phoneMaskPattern: String
    get() {
        if (isBlank()) return UNIVERSAL_MASK
        var current: JSONObject = PhoneMasks.VALUE_DEFAULT_VALUE
        var countryCodeInd = 0
        while (!current.has("value")) {
            if (countryCodeInd >= length) {
                current = current.get(COUNTRY_CODE_END_MARKER) as JSONObject
                break
            }
            val digit = get(countryCodeInd++).toString()
            current = current.get(
                if (current.has(digit)) digit else COUNTRY_CODE_END_MARKER
            ) as JSONObject
        }
        return current.getString("value") + PhoneMasks.EXTRA_NUMBERS
    }

File: client/web/divkit/src/utils/mask/phoneInputMask.ts
import type { LogError, WrappedError } from '../wrapError';
import { BaseInputMask, MaskCharDynamic } from './baseInputMask';
import { textDiff } from '../textDiff';
import data from '../../../../../../shared_data/phone-masks.json';

const UNIVERSAL_MASK = '000000000000000';
const COUNTRY_CODE_END_MARKER = '*';
const EXTRA_NUMBERS = '00';

const DEFAULT_DECODING = [{
    key: '0',
    filter: '\\d',
    placeholder: '_'
}];

export class PhoneInputMask extends BaseInputMask {
    protected decimalSeparator = '.';
    protected localeDigits: Record<string, string> = {};
    protected trimZeroRegExp = new RegExp('');

    constructor(private readonly logError: LogError) {
        super({
            pattern: phoneMaskPattern(''),
            decoding: DEFAULT_DECODING,
            alwaysVisible: false
        });
    }

    override overrideRawValue(newRawValue: string): void {
        this.tryInvalidateMaskDataWith(newRawValue);
        super.overrideRawValue(newRawValue);
    }

    override applyChangeFrom(newValue: string, position?: number): void {
        const diff = textDiff(this.value, newValue);

        if (position !== undefined) {
            diff.start = Math.max(0, position - diff.added);
        }

        const oldRawValue = this.rawValue;

        const tailStart = this.replaceBodyTail(diff, newValue);

        const newRawValue = this.rawValue;
        const newPattern = this.newMaskPatternFor(newRawValue);

        if (newPattern == null) {
            this.calculateCursorPosition(diff, tailStart);
            return;
        }

        this.updateMaskDataWith(newPattern);
        this.replaceChars(newRawValue, 0);

        const rawValueDiff = textDiff(oldRawValue, newRawValue);
        const dynamicDestination = rawValueDiff.start + rawValueDiff.added;
        this.calculateCursorPositionBy(dynamicDestination);
    }

    private calculateCursorPositionBy(dynamicDestination: number): void {
        let index = 0;
        let dynamicCounter = 0;

        while (index < this.destructedValue.length && dynamicCounter < dynamicDestination) {
            if (this.destructedValue[index++] instanceof MaskCharDynamic) {
                dynamicCounter++;
            }
        }

        this.cursorPos = this.firstHolderAfter(index);
    }

    private tryInvalidateMaskDataWith(rawValue: string): void {
        const newPattern = this.newMaskPatternFor(rawValue);
        if (newPattern) {
            this.updateMaskDataWith(newPattern);
        }
    }

    private newMaskPatternFor(rawValue: string): string | null {
        const newPattern = phoneMaskPattern(rawValue);
        const currentPattern = this.maskData.pattern;
        return newPattern !== currentPattern ? newPattern : null;
    }

    private updateMaskDataWith(newPattern: string) {
        return this.updateMaskData({
            pattern: newPattern,
            decoding: DEFAULT_DECODING,
            alwaysVisible: this.maskData.alwaysVisible
        }, false);
    }

    onException(exception: WrappedError) {
        this.logError(exception);
    }
}

function resolveObject(obj: any) {
    if ('$ref' in obj) {
        return data.constants[obj.$ref.split('/').pop() as keyof typeof data.constants];
    }
    return obj;
}

function phoneMaskPattern(val: string): string {
    if (!val) {
        return UNIVERSAL_MASK;
    }
    let current: any = data.properties.value.default_value;
    let countryCodeInd = 0;
    while (!('value' in current)) {
        if (countryCodeInd >= val.length) {
            current = resolveObject(current[COUNTRY_CODE_END_MARKER]);
            break;
        }
        const digit = val[countryCodeInd++];
        current = resolveObject(current[digit in current ? digit : COUNTRY_CODE_END_MARKER]);
    }
    return current.value + EXTRA_NUMBERS;
}
